File: ecoflow_logger/metrics.py
"""Extracts the metrics we care about from a raw EcoFlow quota payload.

The DELTA 2 Max reports dozens of quota fields; the exact set of keys
that are populated can vary slightly by firmware version. Each metric
below is looked up via an ordered list of candidate keys, and the first
one present in the payload wins.
"""

from __future__ import annotations

from typing import Any, NamedTuple

# State of charge, in percent. `pd.soc` confirmed against a real DELTA 2
# Max (2026-09-20); the bmsMaster.* keys were an unverified guess that
# never matched any real device response and are kept as fallbacks for
# other EcoFlow models that may use them.
SOC_KEYS = [
    "pd.soc",
    "bms_emsStatus.lcdShowSoc",
    "bms_bmsStatus.soc",
    "bmsMaster.soc",
    "bmsMaster.f32ShowSoc",
    "soc",
]

# Total power flowing into the device, in watts (AC + solar + car input).
WATTS_IN_KEYS = ["pd.wattsInSum", "inv.inputWatts"]

# Total power flowing out of the device, in watts (AC + USB + DC output).
WATTS_OUT_KEYS = ["pd.wattsOutSum", "inv.outputWatts"]

# Extra/expansion battery pack ("Smart Extra Battery"), reported by
# EcoFlow as a "slave" BMS pack. bms_slave_bmsSlaveStatus_1.* and
# pd.bpPowerSoc were both confirmed present in a real DELTA 2 Max +
# extra-battery quota response (2026-09-20) -- unverified which one
# actually holds the live value, so both are tried.
EXTRA_BATTERY_SOC_KEYS = [
    "bms_slave_bmsSlaveStatus_1.soc",
    "bms_slave_bmsSlaveStatus_1.f32ShowSoc",
    "pd.bpPowerSoc",
]
EXTRA_BATTERY_WATTS_IN_KEYS = ["bms_slave_bmsSlaveStatus_1.inputWatts"]
EXTRA_BATTERY_WATTS_OUT_KEYS = ["bms_slave_bmsSlaveStatus_1.outputWatts"]

# Individual solar (PV) input channels -- the DELTA 2 Max has two separate
# MPPT inputs, shown as two "Solar" tiles in the app. mppt.inWatts /
# mppt.pv2InWatts confirmed present in a real quota response (2026-09-20);
# pd.pv1ChargeWatts/pd.pv2ChargeWatts are the pd-namespace mirrors seen for
# other metrics (e.g. pd.wattsInSum vs inv.inputWatts) and kept as fallbacks.
PV1_WATTS_KEYS = ["mppt.inWatts", "pd.pv1ChargeWatts"]
PV2_WATTS_KEYS = ["mppt.pv2InWatts", "pd.pv2ChargeWatts"]


class Reading(NamedTuple):
    soc_percent: float | None
    watts_in: float | None
    watts_out: float | None
    extra_battery_soc_percent: float | None
    extra_battery_watts_in: float | None
    extra_battery_watts_out: float | None
    pv1_watts: float | None
    pv2_watts: float | None


def _first_present(data: dict[str, Any], keys: list[str]) -> float | None:
    for key in keys:
        if key in data and data[key] is not None:
            try:
                return float(data[key])
            except (TypeError, ValueError):
                continue
    return None


def extract_reading(quota_data: dict[str, Any]) -> Reading:
    """Pull SoC and power in/out out of a raw ``get_all_quota()`` payload."""
    return Reading(
        soc_percent=_first_present(quota_data, SOC_KEYS),
        watts_in=_first_present(quota_data, WATTS_IN_KEYS),
        watts_out=_first_present(quota_data, WATTS_OUT_KEYS),
        extra_battery_soc_percent=_first_present(quota_data, EXTRA_BATTERY_SOC_KEYS),
        extra_battery_watts_in=_first_present(quota_data, EXTRA_BATTERY_WATTS_IN_KEYS),
        extra_battery_watts_out=_first_present(quota_data, EXTRA_BATTERY_WATTS_OUT_KEYS),
        pv1_watts=_first_present(quota_data, PV1_WATTS_KEYS),
        pv2_watts=_first_present(quota_data, PV2_WATTS_KEYS),
    )

File: android/core/src/main/kotlin/com/ecoflowlogger/core/Metrics.kt
package com.ecoflowlogger.core

/**
 * Extracts the metrics this app cares about from a raw EcoFlow quota
 * payload. Ported field-for-field from ecoflow_logger/metrics.py, which
 * carries the verification history for each key list below (see that
 * file's comments and git history for the real-device evidence).
 */
object Metrics {

    val SOC_KEYS = listOf(
        "pd.soc",
        "bms_emsStatus.lcdShowSoc",
        "bms_bmsStatus.soc",
        "bmsMaster.soc",
        "bmsMaster.f32ShowSoc",
        "soc",
    )

    val WATTS_IN_KEYS = listOf("pd.wattsInSum", "inv.inputWatts")
    val WATTS_OUT_KEYS = listOf("pd.wattsOutSum", "inv.outputWatts")

    private const val KIT_INFO_KEY = "bms_kitInfo.watts"

    // Only consulted when the device reports no kit info to resolve the
    // attached slot from (see attachedExtraBatterySlot). A real DELTA 2 Max
    // was observed reporting a stale bms_slave_bmsSlaveStatus_1 block
    // (frozen at a wrong SoC/wattage for hours) alongside a correct
    // bms_slave_bmsSlaveStatus_2 block for the pack that was actually
    // attached -- both carrying the same packSn. Assuming slot 1 is wrong
    // often enough that it's a last resort, not a primary key list.
    val FALLBACK_EXTRA_BATTERY_SOC_KEYS = listOf(
        "bms_slave_bmsSlaveStatus_1.soc",
        "bms_slave_bmsSlaveStatus_1.f32ShowSoc",
        "pd.bpPowerSoc",
    )
    val FALLBACK_EXTRA_BATTERY_WATTS_IN_KEYS = listOf("bms_slave_bmsSlaveStatus_1.inputWatts")
    val FALLBACK_EXTRA_BATTERY_WATTS_OUT_KEYS = listOf("bms_slave_bmsSlaveStatus_1.outputWatts")

    val PV1_WATTS_KEYS = listOf("mppt.inWatts", "pd.pv1ChargeWatts")
    val PV2_WATTS_KEYS = listOf("mppt.pv2InWatts", "pd.pv2ChargeWatts")

    private fun toDoubleOrNull(value: Any?): Double? = when (value) {
        is Number -> value.toDouble()
        is String -> value.toDoubleOrNull()
        else -> null
    }

    private fun firstPresent(data: Map<String, Any?>, keys: List<String>): Double? {
        for (key in keys) {
            if (!data.containsKey(key)) continue
            val num = toDoubleOrNull(data[key]) ?: continue
            return num
        }
        return null
    }

    /**
     * Returns the slave-slot number (1-based, matching the
     * bms_slave_bmsSlaveStatus_N key naming) of the extra battery that is
     * actually attached, per bms_kitInfo.watts' 0-based list -- the
     * authoritative slot list, where the attached pack is flagged
     * avaFlag == 1. Returns null if the device reports no kit info at all.
     */
    private fun attachedExtraBatterySlot(data: Map<String, Any?>): Int? {
        val kits = data[KIT_INFO_KEY] as? List<*> ?: return null
        for ((index, kit) in kits.withIndex()) {
            val kitMap = kit as? Map<*, *> ?: continue
            if (toDoubleOrNull(kitMap["avaFlag"]) == 1.0) return index + 1
        }
        return null
    }

    private data class ExtraBatteryValues(val soc: Double?, val wattsIn: Double?, val wattsOut: Double?)

    private fun extraBatteryValues(data: Map<String, Any?>): ExtraBatteryValues {
        val slot = attachedExtraBatterySlot(data)
            ?: return ExtraBatteryValues(
                soc = firstPresent(data, FALLBACK_EXTRA_BATTERY_SOC_KEYS),
                wattsIn = firstPresent(data, FALLBACK_EXTRA_BATTERY_WATTS_IN_KEYS),
                wattsOut = firstPresent(data, FALLBACK_EXTRA_BATTERY_WATTS_OUT_KEYS),
            )
        val prefix = "bms_slave_bmsSlaveStatus_$slot."
        return ExtraBatteryValues(
            soc = firstPresent(data, listOf("${prefix}soc", "${prefix}f32ShowSoc")),
            wattsIn = firstPresent(data, listOf("${prefix}inputWatts")),
            wattsOut = firstPresent(data, listOf("${prefix}outputWatts")),
        )
    }

    fun extractReading(data: Map<String, Any?>): Reading {
        val extra = extraBatteryValues(data)
        return Reading(
            socPercent = firstPresent(data, SOC_KEYS),
            wattsIn = firstPresent(data, WATTS_IN_KEYS),
            wattsOut = firstPresent(data, WATTS_OUT_KEYS),
            extraBatterySocPercent = extra.soc,
            extraBatteryWattsIn = extra.wattsIn,
            extraBatteryWattsOut = extra.wattsOut,
            pv1Watts = firstPresent(data, PV1_WATTS_KEYS),
            pv2Watts = firstPresent(data, PV2_WATTS_KEYS),
        )
    }
}
